Add the NuGet SPDX License Model

nkolev92 opened the pull request:
## Bug
Fixes: https://github.com/NuGet/Home/issues/7301  
Regression: No  
If Regression then when did it last work:   
If Regression then how are we preventing it in future:   

## Fix
Details: 
Technical details can be found in https://github.com/NuGet/Engineering/pull/1704/. That spec will be made public when the details are ironed out.

This PR solely concerns itself with defining the NuGet APIs for SPDX-like license expressions and the parsing behavior.

Note that the tokenizers are not the most performant ones. They are focusing on correctness.
Follow up to improve the perf is [here](https://github.com/NuGet/Home/issues/7302). I wanted to focus on correctness first, and don't really want to made the PR too large. 

The class diagram is as below: 
The green arrows are inheritance. 
The grey ones are composition. 

NuGetLicenseExpression is the root type. 
![image](https://user-images.githubusercontent.com/2878341/46064731-0f554f80-c125-11e8-8097-edd214e15bf6.png)

## Testing/Validation
Tests Added: Yes 
Reason for not adding tests:  
Validation done: Manual,

nkolev92 committed:
Add SPDX License expression models and parsers

nkolev92 committed:
remove infix to postfix conversion algorithm

nkolev92 committed:
Eliminate the need for an extra stack

nkolev92 committed:
small refactoring

nkolev92 committed:
Cleanup

nkolev92 committed:
work

nkolev92 committed:
add skeleton for NuGetLicenseData class

nkolev92 committed:
add license data class

nkolev92 committed:
Add code generation for NuGet License Data

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
update the licenses

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
work

nkolev92 committed:
Revert unnecesary changes

nkolev92 committed:
Refactoring and cleanup

loic-sharma commented on test/TestExtensions/GenerateLicenseList/Program.cs:
```diff
@@ -0,0 +1,56 @@
+            if (args.Count() != 3)
+            {
+                Console.WriteLine("This tool expects 3 arguments: licenses.json, exceptions.json, targetFileLocation");
+                return -1;
```
I would add a note that this tool should be run using `UpdateNuGetLicenseSPDXList.ps1`

nkolev92 committed:
work

loic-sharma commented on test/TestExtensions/GenerateLicenseList/Program.cs:
```diff
@@ -0,0 +1,56 @@
+                    var issue = $"ID: {codeIssue.Id}, Message: {codeIssue.GetMessage()}, Location: {codeIssue.Location.GetLineSpan()}, Severity: {codeIssue.Severity}";
+                    Console.WriteLine(issue);
+                    Console.WriteLine("License List file generation failed.");
+                }
```
I'd move this line out of the `foreach`

nkolev92 replied:
Nice catch :)

loic-sharma commented on src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionTokenizer.cs:
```diff
@@ -0,0 +1,116 @@
+        }
+
+        /// <summary>
+        /// The valid characters for a license identifier are a-zA-Z0-9.-+
```
>a-zA-Z0-9.-+ [](start = 62, length = 12)

Why didn't you use this regex expression?

nkolev92 replied:
Perf concern(garbage and class load).
This tokenizer is very dumb and I'm changing a bit about it right now. 
It was something whipped up in 1 hour to just unblock myself :)

loic-sharma replied:
My guess is that regex will be faster than a naive tokenizer. This may be worth a mini benchmark 😄

nkolev92 replied:
It's actually rather difficult to write a regex to parse out the tokens for a valid license expression. 
It also destroys the debug-ability. 

Using regex to find invalid characters is a bit of an overkill, and we can't really do better than `O(n)` where n is the char count.

loic-sharma replied:
To clarify, I'm only suggesting regex to simplify `HasValidCharacters` - not the rest of the tokenizer.

nkolev92 replied:
Ahem, clear. 
Then the 2nd part applies :) 

regex's best time complexity `O(n)` + space for the construction.

loic-sharma replied:
Regex engines have a few tricks up their sleeves (clever algorithms, SIMD, etc... See [this article by ripgrep](https://blog.burntsushi.net/ripgrep/#literal-optimizations)). If performance is the only concern, I think regex may be a viable alternative.

That being said, I only suggested regex to simplify the code.

nkolev92 replied:
Sounds good. 
Not too set on it either way. 
Thoughts @NuGet/nuget-client

jainaashish replied:
if it's only to check if `_value` has all the valid characters then I'd also suggest for regex. It will be way simplified in couple of lines + no worse performance what you've currently...

nkolev92 replied:
I don't care too much either way, I can change it, just don't see the point of jitting regex classes for a very dumb check.

loic-sharma commented on src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionTokenizer.cs:
```diff
@@ -0,0 +1,116 @@
+        internal IEnumerable<LicenseExpressionToken> Tokenize()
+        {
+            var potentialTokens = _value.Split(' ');
+            foreach (var token in potentialTokens)
```
Should you use `StringSplitOptions.RremoveEmptyEntries` here? Line 71 might throw if I put two spaces in between tokens.

nkolev92 replied:
It won't throw as I have a length check (and tests) for it, but that's a great catch. 
Slipped my mind about how this split method worked.

loic-sharma commented on src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionParser.cs:
```diff
@@ -0,0 +1,211 @@
+                switch (token.TokenType)
+                {
+                    case LicenseTokenType.OPENING_BRACKET:
+                        if (!firstPass && !token.TokenType.IsValidPrecedingToken(lastTokenType))
```
>IsValidPrecedingToken [](start = 59, length = 21)

What do you think of `IsValidPrecedingToken` accepting a null `LicenseTokenType` to represent that this is the first pass? This may simplify the parsing logic

Also, could `IsValidPrecedingToken` throw the argument exception if the tokens aren't valid?

nkolev92 replied:
LicenseTokenType is an enum, so usually not a fan of nullable enums. Happy to chat through that though. 

I think it's more readable if the exception is thrown from the parser rather than some generic method.

loic-sharma commented on src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionParser.cs:
```diff
@@ -0,0 +1,211 @@
+                        break;
+
+                    case LicenseTokenType.VALUE:
+                        if (!firstPass && !token.TokenType.IsValidPrecedingToken(lastTokenType))
```
What do you think of moving this case to the top of the switch for readability's sake? It took me a minute to realize the difference between the `operandStack` and the `operatorStack`. A comment for those two variables would also be helpful.

nkolev92 replied:
Moved it. 

The operand/operator stack should be self descriptive though? 

I will add some more comments about how the algorithm works overall and ping you.

loic-sharma commented on src/NuGet.Core/NuGet.Packaging/Licenses/LicenseTokenType.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace NuGet.Packaging.Licenses
+{
+    internal enum LicenseTokenType
```
>LicenseTokenType [](start = 18, length = 16)

I would add a comment that these are ordered by priority

nkolev92 committed:
rework the tokenizer

nkolev92 committed:
work

nkolev92 committed:
address comments

nkolev92 committed:
cleanup

nkolev92 commented:
//cc @NuGet/nuget-client This is ready for review for now. 
@agr @joelverhagen @loic-sharma would love your input on the APIs as you'd be using the same ones in the gallery validations.

rohit21agrawal commented on scripts/utils/UpdateNuGetLicenseSPDXList.ps1:
```diff
@@ -0,0 +1,30 @@
+$licenseListDirectory = $([System.IO.Path]::Combine($env:TEMP, "NuGet", "licenseList"))
+
+try {
+    $gitRepo = "https://github.com/spdx/license-list-data.git"
```
if you just need a few selected files, it might be better to use GitHub's rest API to download specific files from repositories instead of cloning the entire thing.

nkolev92 replied:
Fixed.

rohit21agrawal commented on src/NuGet.Core/NuGet.Packaging/Licenses/LicenseTokenType.cs:
```diff
@@ -0,0 +1,18 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace NuGet.Packaging.Licenses
```
if you are introducing this namespace, the nput LicenseOperator  and LicenseExpressionType in this too

nkolev92 replied:
I missed that somehow. 
That namespace shouldn't exist. Will give it another pass to make sure the namespaces are right.

nkolev92 committed:
fix namespaces

nkolev92 committed:
fix the imports

nkolev92 committed:
fix internals visible to

loic-sharma commented on src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionParser.cs:
```diff
@@ -0,0 +1,217 @@
+        /// <param name="infixTokens"></param>
+        /// <returns>NuGetLicenseExpression</returns>
+        /// <exception cref="ArgumentException">If the expression has invalid characters.</exception>
+        /// <exception cref="ArgumentException">If the expression itself is invalid. Example: MIT OR OR Apache-2.0, or the MIT or Apache-2.0, because the expressions are case sensitive.</exception>
```
Is this list supposed to be exhaustive? If so, `LicenseExpressionTokenizer` throws an `ArgumentNullException` if the string is null or whitespace.

Also, I think the Gallery will want a non-throwing API to verify that the expression is valid. Something like `public static bool TryParse(string expression, out NuGetLicenseExpression licenseExpression)`

nkolev92 replied:
It's not exhaustive but maybe it should be. 
I will try to add all the cases there. 

Regarding the 2nd part. That should be easy to add, although I think you can/should use the messages thrown by the APIs to display to the end-user.

loic-sharma replied:
That's a good point. Maybe instead of throwing you could return a result object that contains a result "code" (Accepted, MismatchedParentheses, etc..) and the parsed `LicenseExpression`, if any. The gallery uses this pattern for validation of uploaded nupkgs (see [this type](https://github.com/NuGet/NuGetGallery/blob/master/src/NuGetGallery/Services/PackageValidationResult.cs) used by [this validation code](https://github.com/NuGet/NuGetGallery/blob/master/src/NuGetGallery/Services/PackageUploadService.cs#L43)), and the client uses this pattern for [signature verification](https://github.com/NuGet/NuGet.Client/blob/3803820961f4d61c06d07b179dab1d0439ec0d91/src/NuGet.Core/NuGet.Packaging/Signing/Verification/VerifySignaturesResult.cs).

I think using the result pattern fits well here as there's "expected" failures that consumers should properly handle.

nkolev92 replied:
I don't think that pattern is appropriate for a parser. 
I wouldn't mind that approach if this was a validator, but this is closer to what we do in NuGetVersion.Parse than anything else. 

For example, we will want to warn about certain things(non standard licenses) which are relevant only to the consumer of the APIs, not the parser itself.

agr commented on src/NuGet.Core/NuGet.Packaging/Licenses/LicenseTokenType.cs:
```diff
@@ -0,0 +1,18 @@
+        OR,
+        OPENING_BRACKET,
+        CLOSING_BRACKET,
+        VALUE,
```
`VALUE` sounds a bit vague. `IDENTIFIER` or even `LICENSE_IDENTIFIER`?
And there is no separation between "License" and "Exception" planned, right?

nkolev92 replied:
This is internal and only for the tokenizer. 

At that point of time, we don't know if it's a license or an exception. 

identifier is not bad.

PatoBeltran replied:
I agree with @agr, `IDENTIFIER` might be a better name.

agr commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseException.cs:
```diff
@@ -0,0 +1,52 @@
+
+namespace NuGet.Packaging
+{
+    public class NuGetLicenseException
```
Missing documentation

nkolev92 commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseException.cs:
```diff
@@ -0,0 +1,52 @@
+    {
+        public string Identifier { get; }
+
+        public bool IsStandardException { get; }
```
This is redundant now, somehow I missed removing it.

agr commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicense.cs:
```diff
@@ -0,0 +1,120 @@
+        /// </summary>
+        public bool IsStandardLicense { get; }
+
+        public NuGetLicense(string identifier, bool plus, bool isStandardLicense)
```
Should it be private? It seems to allow creating some invalid combinations of `Identifier`, `Plus` and `IsStandardLicense` properties.

nkolev92 replied:
Yeah, it should be private.

agr commented on test/NuGet.Core.Tests/NuGet.Packaging.Test/LicensesTests/LicenseExpressionTokenizerTests.cs:
```diff
@@ -0,0 +1,119 @@
+        [InlineData("aND", "VALUE")]
+        public void LicenseExpressionTokenizer_TokenizesSimpleExpressionCorrectly(string value, string type)
+        {
+            Enum.TryParse(type, out LicenseTokenType tokenType);
```
Why not have `tokenType` as a method argument instead of string parsing?

nkolev92 replied:
Because the type is internal (exposed for tests via internalsvisibleto). However xunit needs the parameter types to be public.

nkolev92 committed:
address PR feedback

PatoBeltran commented on scripts/utils/UpdateNuGetLicenseSPDXList.ps1:
```diff
@@ -0,0 +1,30 @@
+try {
+    $gitRepo = "https://github.com/spdx/license-list-data.git"
+    Write-Host "Downloading the license list from $gitRepo"
+    git clone -b master --single-branch https://github.com/spdx/license-list-data.git $licenseListDirectory
```
why not use `$gitRepo` here?

nkolev92 replied:
You were on an old version when looking at this :) 
I changed it around based on Rohit's comments.

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionParser.cs:
```diff
@@ -0,0 +1,217 @@
+        /// Later the postfix expression is evaluated into an object model that represents the expression. Note that brackets are dropped in this conversion and this is not round-trippable.
+        /// The token precedence helps make sure that the expression is a valid infix one. 
+        /// </summary>
+        /// <param name="infixTokens"></param>
```
Remove `params` tag for this parameter that doesn't exist and add one for `expression`

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionParser.cs:
```diff
@@ -0,0 +1,217 @@
+        /// The token precedence helps make sure that the expression is a valid infix one. 
+        /// </summary>
+        /// <param name="infixTokens"></param>
+        /// <returns>NuGetLicenseExpression</returns>
```
should you add more description to what the return value is? The type doesn't tell you much as the function already gives you a return type.

nkolev92 replied:
I think the descriptions in the license expression type and it's members should be the place where the user learns about it. 

I will update those ones.

nkolev92 replied:
I have added docs for all types and did some renaming.

nkolev92 committed:
Don't download the full repo

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionToken.cs:
```diff
@@ -0,0 +1,31 @@
+
+        public override string ToString()
+        {
+            return $"Value:{Value}, Type: {TokenType}";
```
add space after `Value: `

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionTokenizer.cs:
```diff
@@ -0,0 +1,132 @@
+            }
+            return null;
+        }
+
```
Maybe add a comment to the method `<remarks>` saying that the method assumed brackets have already been parsed.

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/LogicalOperator.cs:
```diff
@@ -0,0 +1,26 @@
+{
+    public class LogicalOperator : LicenseOperator
+    {
+        public LogicalOperator(LogicalOperatorType logicalOperatorType, NuGetLicenseExpression left, NuGetLicenseExpression right) : base()
```
can `left` and `right` be null?

Also, add a new line for `: base()`

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicense.cs:
```diff
@@ -0,0 +1,123 @@
+        /// <summary>
+        /// Parse a licenseIdentifier. If a licenseIdentifier is deprecated, this will throw. Non-standard licenses get parsed into a object model as well.
+        /// </summary>
+        /// <param name="licenseIdentifier"></param>
```
add description to empty `param` tag

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicense.cs:
```diff
@@ -0,0 +1,123 @@
+        /// Parse a licenseIdentifier. If a licenseIdentifier is deprecated, this will throw. Non-standard licenses get parsed into a object model as well.
+        /// </summary>
+        /// <param name="licenseIdentifier"></param>
+        /// <returns></returns>
```
Add description or remove empty `returns` tag

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicense.cs:
```diff
@@ -0,0 +1,123 @@
+        /// <param name="licenseIdentifier"></param>
+        /// <returns></returns>
+        /// <exception cref="ArgumentException">If the identifier is deprecated</exception>
+        /// <exception cref="ArgumentException">If it's null or empty.</exception>
```
use `paramref` to refer to `licenseIdentifier`

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicense.cs:
```diff
@@ -0,0 +1,123 @@
+                }
+            }
+            // This will not happen in production code as the tokenizer takes cares of that. 
+            throw new ArgumentException(Strings.ArgumentCannotBeNullOrEmpty, nameof(licenseIdentifier));
```
Add this exception to the exceptions list in the comments

nkolev92 replied:
Cleaned up the comments. 
ArgumentException is already in the lsit :)

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseData.cs:
```diff
@@ -0,0 +1,486 @@
+    {
+        public LicenseData(string licenseID, int referenceNumber, bool isOsiApproved, bool isDeprecatedLicenseId)
+        {
+            LicenseID = licenseID;
```
can `licenseID` be `null`?

nkolev92 replied:
This is autogened. 
Will never happen.

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseData.cs:
```diff
@@ -0,0 +1,486 @@
+    {
+        public ExceptionData(string licenseID, int referenceNumber, bool isDeprecatedLicenseId)
+        {
+            LicenseExceptionID = licenseID;
```
can `licenseID` be `null`?

nkolev92 replied:
Same as above. It's auto gened.

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseException.cs:
```diff
@@ -0,0 +1,55 @@
+            Identifier = identifier ?? throw new ArgumentNullException(nameof(identifier));
+        }
+
+        public static NuGetLicenseException Parse(string exceptionIdentifier)
```
Add comments

PatoBeltran reviewed (commented):
It looks good overall, only small comments

nkolev92 committed:
Clean up the APIs, add comments

nkolev92 committed:
address comments

nkolev92 committed:
add more comments

nkolev92 committed:
add more comments

nkolev92 commented:
Addressed all the comments 🔔

nkolev92 committed:
add missed out file

joelverhagen commented on src/NuGet.Core/NuGet.Protocol/LocalRepositories/LocalV2FindPackageByIdResource.cs:
```diff
@@ -313,6 +313,7 @@ private IReadOnlyList<LocalPackageInfo> GetPackageInfosCore(string id, ILogger l
                     try
                     {
                         reader = new NuspecReader(packageReader.GetNuspec());
+                        // similarly to this, I will need to add a way to get the License file...probably as a stream.
```
>similarly [](start = 27, length = 9)

TODOs should reference a GitHub issue.

nkolev92 replied:
Actually this should be reverted :D

joelverhagen commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseOperator.cs:
```diff
@@ -0,0 +1,25 @@
+        /// LogicalOperator means it's AND or OR and <see cref="NuGetLicenseLogicalOperator"/>
+        /// NuGetLicenseWithOperator means it's the WITH operator and <see cref="WithOperator"/>
+        /// </summary>
+        public NuGetLicenseOperatorType OperatorType { get; protected set; }
```
>protected [](start = 60, length = 9)

This could be made `{ get; }` if you pass the values into the base constructor right? Then we know this is a truly immutable type.

nkolev92 committed:
revert unnecesary change

nkolev92 commented:
I would to get people's opinion on having a new namespace vs prefixing. 

I double checked...and we have NuGet.Packaging.Signing, and I don't think it's a bad idea to have NuGet.Packaging.Licenses

joelverhagen commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseData.cs:
```diff
@@ -0,0 +1,486 @@
+
+        public static Dictionary<string, LicenseData> LicenseList = new Dictionary<string, LicenseData>()
+        {
+            {"0BSD", new LicenseData(licenseID: "0BSD", referenceNumber: 1, isOsiApproved: false, isDeprecatedLicenseId: false) },
```
nit: space between `{` and `"` since you have one before `}`

joelverhagen commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseData.cs:
```diff
@@ -0,0 +1,486 @@
+            {"Nunit", new LicenseData(licenseID: "Nunit", referenceNumber: 373, isOsiApproved: false, isDeprecatedLicenseId: true) },
+            {"StandardML-NJ", new LicenseData(licenseID: "StandardML-NJ", referenceNumber: 374, isOsiApproved: false, isDeprecatedLicenseId: true) },
+            {"eCos-2.0", new LicenseData(licenseID: "eCos-2.0", referenceNumber: 375, isOsiApproved: false, isDeprecatedLicenseId: true) },
+            {"wxWindows", new LicenseData(licenseID: "wxWindows", referenceNumber: 376, isOsiApproved: false, isDeprecatedLicenseId: true) },
```
nit: can these be sorted alphabetically?

nkolev92 replied:
I prefer to keep the order the same as they are in https://github.com/spdx/license-list-data/blob/master/json/licenses.json. 

They are sorted by reference number.

joelverhagen commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseExpressionExtensions.cs:
```diff
@@ -0,0 +1,83 @@
+                    switch (licenseOperator.OperatorType)
+                    {
+                        case NuGetLicenseOperatorType.LogicalOperator:
+                            var logicalOperator = expression as NuGetLicenseLogicalOperator;
```
>as [](start = 61, length = 2)

Why `as`? Why not cast?

nkolev92 replied:
habit...as can be used everywhere where casting is, but not the other way around :) 

Fixed it.

joelverhagen commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseExpressionExtensions.cs:
```diff
@@ -0,0 +1,83 @@
+        /// <param name="expression"></param>
+        /// <param name="licenseProcessor"></param>
+        /// <param name="exceptionProcessor"></param>
+        public static void OnEachLeafNode(this NuGetLicenseExpression expression, Action<NuGetLicense> licenseProcessor, Action<NuGetLicenseException> exceptionProcessor)
```
>OnEachLeafNode [](start = 27, length = 14)

Perhaps we should document this as pre-order traversal? That's what it is, right?

nkolev92 replied:
Not sure we can call it either since we don't really act on the current node :)

joelverhagen commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseLogicalOperator.cs:
```diff
@@ -0,0 +1,35 @@
+
+        public override string ToString()
+        {
+            return $"{Left.ToString()} {LogicalOperatorType} {Right.ToString()}";
```
This is round-trippable if you put in parenthesis right? Might be nice, haha.

nkolev92 replied:
Yeah potentially :)
I'd still lose formatting at least.

joelverhagen commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseData.cs:
```diff
@@ -0,0 +1,486 @@
+    {
+        public static string LicenseListVersion = "v3.1-67-geb2589b";
+
+        public static Dictionary<string, LicenseData> LicenseList = new Dictionary<string, LicenseData>()
```
>Dictionary [](start = 22, length = 10)

nit: IReadOnlyDictionary

joelverhagen replied:
And `readonly`

---
In reply to: [218980640](https://github.com/NuGet/NuGet.Client/pull/2424#discussion_r218980640) [](ancestors = 218980640)

joelverhagen commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseLogicalOperatorType.cs:
```diff
@@ -0,0 +1,14 @@
+    /// </summary>
+    public enum NuGetLicenseLogicalOperatorType
+    {
+        AND,
```
>AND [](start = 8, length = 3)

nit: This casing is inconsistent with the rest of the enums AFAIK. `And` and `Or` would be more consistent. Note that if you go this route the `ToString` of `LicenseOperator` will look weird so maybe fix that up too if you make this change.

nkolev92 replied:
Yeah get your point. 
Fixed. 
I'd rather have the types be consistent.

nkolev92 committed:
add a failing test

nkolev92 committed:
test

nkolev92 committed:
add a namespace, rework the names

nkolev92 committed:
address feedback

nkolev92 committed:
address feedback about the license data container

nkolev92 commented:
@joelverhagen addressed your feedback. 

I'd love some feedback on the naming too. 

I also have an expression that generates the wrong object model:
`(AMDPLPA AND BSD-2-Clause) AND (MIT OR Apache-2.0)`
but whether an expression is valid still correct. We never really use the object model for much. 

I will fix it, but the rest of the work depends on this being checked in first, so I'd rather do that in an additional PR. 

**EDIT**:
It's been fixed.

nkolev92 committed:
rework the expression generation

nkolev92 committed:
work

nkolev92 commented:
Update; 
The above mentioned bug has been fixed.

nkolev92 committed:
cleanup

nkolev92 committed:
work

jainaashish commented on scripts/utils/UpdateNuGetLicenseSPDXList.ps1:
```diff
@@ -0,0 +1,62 @@
+    $exceptionsFileName = "exceptions.json"
+
+    DownloadFileFromUrl -BaseUrl $licenseListBaseUrl -Directory $licenseListDirectory -FileName $licenseFileName
+    DownloadFileFromUrl -BaseUrl $licenseListBaseUrl -Directory $licenseListDirectory -FileName $exceptionsFileName
```
why downloading the same json in 2 different files?

nkolev92 replied:
I'm not downloading the same json. 
They have the same base url. 
Look at the function https://github.com/NuGet/NuGet.Client/pull/2424/files#diff-f506fa6a1942491a5b15c4ab6335a84bR21
It appends the file name to the url and downloads it.

jainaashish commented on src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionToken.cs:
```diff
@@ -0,0 +1,31 @@
+        /// The token type
+        /// </summary>
+        internal LicenseTokenType TokenType { get; }
+        /// <summary>
```
nit: space

nkolev92 replied:
fixed.

jainaashish commented on src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionTokenizer.cs:
```diff
@@ -0,0 +1,136 @@
+    internal class LicenseExpressionTokenizer
+    {
+        private readonly string _value;
+        /// <summary>
```
nit: space

nkolev92 replied:
fixed.

jainaashish commented on src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionTokenizer.cs:
```diff
@@ -0,0 +1,136 @@
+        {
+            if (string.IsNullOrWhiteSpace(value))
+            {
+                throw new ArgumentException(nameof(value));
```
how about `ArgumentNullException` which is also consistent across NuGet codebase

nkolev92 replied:
Actually quite the opposite. 

We're probably not strict enough and we have a mixture of both, but we should be using ArgumentException. 

See @dtivel's comment. 
https://github.com/NuGet/NuGet.Client/pull/2370#discussion_r217167772

Also code references:
https://github.com/NuGet/NuGet.Client/blob/e9630fe626237a6788110abbdc5d231c14f0cd57/src/NuGet.Core/NuGet.Packaging/Signing/Signatures/NuGetPackageOwners.cs
https://github.com/NuGet/NuGet.Client/blob/e9630fe626237a6788110abbdc5d231c14f0cd57/src/NuGet.Core/NuGet.Packaging/Signing/Utility/AttributeUtility.cs#L163
https://github.com/NuGet/NuGet.Client/blob/e9630fe626237a6788110abbdc5d231c14f0cd57/src/NuGet.Core/NuGet.Commands/CommandRunners/LocalsCommandRunner.cs#L47

PatoBeltran replied:
Is there any string resource to specify that the argument cannot be null or empty that you can use?

nkolev92 replied:
good point Pato. 
Fixed.

jainaashish commented on src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionTokenizer.cs:
```diff
@@ -0,0 +1,136 @@
+            var potentialTokens = _value.Split(new[] { ' ' }, StringSplitOptions.RemoveEmptyEntries);
+            foreach (var token in potentialTokens)
+            {
+                var processingToken = token;
```
trim the toen before start processing

nkolev92 replied:
I am splitting based on a `' '`. It's trimmed already.

jainaashish commented on src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionTokenizer.cs:
```diff
@@ -0,0 +1,136 @@
+                        tokensAfterValue = new List<LicenseExpressionToken>();
+
+                    }
+                    tokensAfterValue.Add(ParseBracket(processingToken[processingToken.Length - 1]));
```
nit: space

nkolev92 replied:
fixed.

jainaashish commented on src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionTokenizer.cs:
```diff
@@ -0,0 +1,136 @@
+        /// Given a string, tokenizes by space into operators and values. The considered operators are, AND, OR, WITH, (, and ). 
+        /// </summary>
+        /// <returns>Tokens</returns>
+        internal IEnumerable<LicenseExpressionToken> Tokenize()
```
I'm not sure if this will work for all the different expressions. For example I tries to dry run it for `1*(ALPHA AND (MIT OR ASDD+312 WITH gdfgg))` but felt like it would fail at the first token itself.

Apart from that, I think we can simplified the logic using Stack so the way it will work is that
- Before parsing, put opening bracket at the start of the expression and closing bracket at the end.
- then keep pushing every char into stack until you find a closing bracket.
- Once you find a closing bracket, then keep poping from stack until a opening bracket.
- Then parse this token, It will be VALUE/ VALUE1 OR VALUE2/ VALUE1 AND VALUE2, etc...
- repeat until the stack become empty

I'm sure there will be many other ways but I'll imagine using stack will be very convenient here and simplify your logic.

nkolev92 replied:
Why are you running 1*(ALPHA AND (MIT OR ASDD+312 WITH gdfgg))? 

The 1* is invalid. 

Furthermore this is simply the tokenizer. 

It will only create tokens..given a valid expression namely something like 
> (0BSD AND (MIT OR Apache-2.0 WITH 389-exception))
it will generate the list of tokens. 
(, 0BSD, AND, (, MIT, OR, Apache-2.0, WITH, 389-exception, ), )

Furthermore in your example:

The * is illegal. 
The values after the + are illegal. The + operator is only valid at the end of an identifier. 

Regarding the parsing logic, it's similar to what you are articulating and the logic is found is 
https://github.com/NuGet/NuGet.Client/pull/2424/files#diff-0a6246a67c49154a731a0cb6a37218a0R10

jainaashish commented on src/NuGet.Core/NuGet.Packaging/Licenses/LogicalOperator.cs:
```diff
@@ -0,0 +1,34 @@
+            base(LicenseOperatorType.LogicalOperator)
+        {
+            LogicalOperatorType = logicalOperatorType;
+            Left = left ?? throw new ArgumentNullException(nameof(left));
```
leftExpression and rightExpression

nkolev92 replied:
It's unnecessarily verbose that way. 
See example implementations of binary operator models. 
https://docs.microsoft.com/en-us/dotnet/api/system.linq.expressions.binaryexpression.left?view=netframework-4.7.2#System_Linq_Expressions_BinaryExpression_Left

jainaashish replied:
I meant `ArgumentNullException(nameof(left)` looking at a exception `left is null` doesn't make any sense

jainaashish commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicense.cs:
```diff
@@ -0,0 +1,123 @@
+                {
+                    return !licenseData.IsDeprecatedLicenseId ?
+                        new NuGetLicense(licenseIdentifier, plus: false, isStandardLicense: true) :
+                        throw new ArgumentException(string.Format(Strings.NuGetLicenseExpression_DeprecatedIdentifier, licenseIdentifier));
```
How about `InvalidOperationException` instead of `ArgumentException`?

jainaashish commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicense.cs:
```diff
@@ -0,0 +1,123 @@
+        /// <returns>NuGetLicense</returns>
+        /// <exception cref="ArgumentException">If the identifier is deprecated</exception>
+        /// <exception cref="ArgumentException">If it's null or empty.</exception>
+        public static NuGetLicense Parse(string licenseIdentifier)
```
how is it taking care of `*` something like `1*1(MIT)`?

nkolev92 replied:
\* is not a valid characters. 
The tokenizer will fail throw if you pass that  string.

jainaashish commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicense.cs:
```diff
@@ -0,0 +1,123 @@
+        /// </summary>
+        /// <param name="value"></param>
+        /// <returns></returns>
+        private static bool HasValidCharacters(string value)
```
it'll make sense to convert it to regex as well. And if it's similar to LicenseExpressionTokenizer one then better move it to a utility class

nkolev92 replied:
Same as before, don't see the point in jitting regex classes, but will change it. 
The allowed char lists are a little bit different.

jainaashish commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseException.cs:
```diff
@@ -0,0 +1,62 @@
+    /// <summary>
+    /// NuGet's internal representation of a license exception identifier. 
+    /// </summary>
+    public class NuGetLicenseException
```
why there is a separate class for Exception, should that be part of NuGetLicense (as part of some license id)? Would there be a exception without an license id?

nkolev92 replied:
It's not part of a nuget license. 
They are separate. 

LICENSE WITH EXCEPTION is a complex expression by definition. 
Any license X exception combination si acceptable.

You cannot have an exception without a license ID, hence why NuGetLicenseException does not inherit NuGetLicenseExpression.

jainaashish replied:
`You cannot have an exception without a license ID, hence why NuGetLicenseException does not inherit NuGetLicenseExpression.`

That's why I dont understand why there is a class with only exception identifier when it can never happens without a license id... May be we can discuss it offline to get clarity.

nkolev92 replied:
Because it's more extensible for the future. We have an object that represents a exception whatever that may be. Future additions to exceptions (maybe we care about OSI approved licenses/exceptions for example for some feature), would be backwards compatible for code using the parser because they won't change the types of what a WithOperator represents.

jainaashish commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseExpressionParser.cs:
```diff
@@ -0,0 +1,201 @@
+        /// </summary>
+        /// <param name="expression"></param>
+        /// <returns>NuGetLicenseExpression</returns>
+        /// <exception cref="ArgumentException">If the expression has invalid characters</exception>
```
`InvalidOperationException` might make more sense and be consistent with rest of the NuGet

nkolev92 replied:
It's not quite an operation. 
This is simply parsing. 

I agree ArgumentException is not great, but I don't think InvalidOperationException is any better. 
Ideally maybe I should have a LicenseParsingException and just catch and rethrow that in the parsing methods. 

Thoughts?

jainaashish replied:
The way I see, It's still an operation of parsing and validating the expression hence it throws different exception when that fails so `InvalidOperationException` seems reasonable to me. Or else, you can derive a new exception something more meaningful to license like `LicenseParsingException` or `LicenseValidationException`

nkolev92 replied:
Refer to: 
https://docs.microsoft.com/en-us/dotnet/api/system.invalidoperationexception?view=netframework-4.7.2

```
The exception that is thrown when a method call is invalid for the object's current state.
```
and https://docs.microsoft.com/en-us/dotnet/api/system.invalidoperationexception?view=netframework-4.7.2#remarks
```
InvalidOperationException is used in cases when the failure to invoke a method is caused by reasons other than invalid arguments. 
Typically, it is thrown when the state of an object cannot support the method call. 
```

The license parsing exception seems more reasonable.
I can have the parser wrap every exception and rethrow a LicenseParsingException.

jainaashish commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseExpressionParser.cs:
```diff
@@ -0,0 +1,201 @@
+        /// <exception cref="ArgumentException">If the expression's brackets are mismatched.</exception>
+        /// <exception cref="ArgumentException">If the licenseIdentifier is deprecated.</exception>
+        /// <exception cref="ArgumentException">If the exception identifier is deprecated.</exception>
+        public static NuGetLicenseExpression Parse(string expression)
```
not sure why there are separate impl for tokens and then parsing the whole expression which also jut make sure if that's a standard expression or complex one and with valid license identifiers.

I believe the earlier comment and logic to parse this expression should work.

nkolev92 replied:
Refer to the earlier comment. 

Having a tokenizer and lexer/parser is a common pattern for lexical analyzers.

jainaashish commented on src/NuGet.Core/NuGet.Packaging/Strings.resx:
```diff
@@ -718,4 +718,33 @@ Valid from:</comment>
+    <value>Mismatched parentheses in the expression.</value>
+  </data>
+  <data name="NuGetLicenseExpression_InvalidExpression" xml:space="preserve">
+    <value>The license expression is invalid.</value>
```
mention the expression as well...?

nkolev92 replied:
I can add that but it doesn't necessarily add too much, as the license expression is unique per package/nuspec.

jainaashish reviewed (changes requested):
I've some questions and concerns about parsing logic itself but overall looks good and in right direction.

nkolev92 reviewed (commented):
I've answered your comments @jainaashish 

I think you might have misunderstood which expressions are valid, we can sync up offline if you'd rather.

nkolev92 committed:
address feedback

nkolev92 committed:
use regex for legal character validations

PatoBeltran commented on scripts/utils/UpdateNuGetLicenseSPDXList.ps1:
```diff
@@ -0,0 +1,62 @@
+    DownloadFileFromUrl -BaseUrl $licenseListBaseUrl -Directory $licenseListDirectory -FileName $exceptionsFileName
+
+    $targetFile = $([System.IO.Path]::GetFullPath($([System.IO.Path]::Combine($PSScriptRoot, "..", "..", "src\NuGet.Core\NuGet.Packaging\Licenses\NuGetLicenseData.cs"))))
+    $generateLicenseList = $([System.IO.Path]::GetFullPath($([System.IO.Path]::Combine($PSScriptRoot, "..", "..", "test\TestExtensions\GenerateLicenseList\GenerateLicenseList.csproj"))))
```
Should we add a test to make sure this file does not change locations?

nkolev92 replied:
The project file itself?
The script would just fail miserably if it does :) 
I think that's good enough :)

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionToken.cs:
```diff
@@ -0,0 +1,32 @@
+
+        internal LicenseExpressionToken(string value, LicenseTokenType tokenType)
+        {
+            Value = value;
```
can this `value` be `null`? If not, change this to `  Value = value ?? throw new ArgumentNullException(nameof(value));`

nkolev92 replied:
Shouldn't. added a check

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionTokenizer.cs:
```diff
@@ -0,0 +1,123 @@
+        /// The valid characters for a license expression are the above whitespace and ().
+        /// </summary>
+        /// <param name="value"></param>
+        /// <returns></returns>
```
fill out `<param` and `<returns>` tags or remove them if they do not provide useful information.

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionTokenizer.cs:
```diff
@@ -0,0 +1,123 @@
+        /// A tokenizer for a license expression.
+        /// This implementation assumes that the input has been sanitized and that there are no invalid characters.
+        /// </summary>
+        /// <param name="value"></param>
```
fill out `<param` tag or remove it if it does not provide useful information.

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionTokenizer.cs:
```diff
@@ -0,0 +1,123 @@
+        /// <summary>
+        /// Given a string, tokenizes by space into operators and values. The considered operators are, AND, OR, WITH, (, and ). 
+        /// </summary>
+        /// <returns>Tokens</returns>
```
Give a more detailed description than `Tokens` or remove if there's no useful description.

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionTokenizer.cs:
```diff
@@ -0,0 +1,123 @@
+                    if (tokensAfterValue == null)
+                    {
+                        tokensAfterValue = new List<LicenseExpressionToken>();
+
```
nit: you can probably delete this extra line

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionTokenizer.cs:
```diff
@@ -0,0 +1,123 @@
+        /// Parses a token type given a string.
+        /// This method assumes that the brackets have been parsed out. 
+        /// </summary>
+        /// <param name="token"></param>
```
fill out `<param` tag or remove it if it does not provide useful information.

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionTokenizer.cs:
```diff
@@ -0,0 +1,123 @@
+        /// This method assumes that the brackets have been parsed out. 
+        /// </summary>
+        /// <param name="token"></param>
+        /// <returns>LicenseExpressionToken</returns>
```
Add a more detailed description or remove. The type is already given by the function itself, no need to have it in the comments

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionType.cs:
```diff
@@ -0,0 +1,16 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+
```
nit: remove extra line

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/LicenseOperator.cs:
```diff
@@ -0,0 +1,26 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+
```
nit: remove extra line

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicense.cs:
```diff
@@ -0,0 +1,112 @@
+        /// The valid characters for a license identifier are a-zA-Z0-9.-
+        /// This method assumes that the trailing + operator has been stripped out.
+        /// </summary>
+        /// <param name="value"></param>
```
fill out `<param` tag or remove it if it does not provide useful information.

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicense.cs:
```diff
@@ -0,0 +1,112 @@
+        /// This method assumes that the trailing + operator has been stripped out.
+        /// </summary>
+        /// <param name="value"></param>
+        /// <returns></returns>
```
Add a description or remove unnecessary `returns` tag

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseException.cs:
```diff
@@ -0,0 +1,62 @@
+        /// <summary>
+        /// Parse an exceptionIdentifier. If the exceptionIdentifier is deprecated, this will throw. Non-standard exception do not get parsed into an object model.
+        /// </summary>
+        /// <param name="exceptionIdentifier"></param>
```
fill out `<param` tag or remove it if it does not provide useful information.

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseException.cs:
```diff
@@ -0,0 +1,62 @@
+        /// Parse an exceptionIdentifier. If the exceptionIdentifier is deprecated, this will throw. Non-standard exception do not get parsed into an object model.
+        /// </summary>
+        /// <param name="exceptionIdentifier"></param>
+        /// <returns>NuGetLicenseException</returns>
```
Add more detail description to `returns` tag. The return type is already given by the function so no need to have it in the comments.

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseExpressionExtensions.cs:
```diff
@@ -0,0 +1,83 @@
+        /// <summary>
+        /// Determines whether all the licenses and exceptions are not deprecated.
+        /// </summary>
+        /// <param name="expression"></param>
```
fill out `<param` tag or remove it if it does not provide useful information.

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseExpressionExtensions.cs:
```diff
@@ -0,0 +1,83 @@
+        /// Determines whether all the licenses and exceptions are not deprecated.
+        /// </summary>
+        /// <param name="expression"></param>
+        /// <returns></returns>
```
Add description or remove if unnecessary

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseExpressionExtensions.cs:
```diff
@@ -0,0 +1,83 @@
+            switch (expression.Type)
+            {
+                case LicenseExpressionType.License:
+
```
nit: remove extra line

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseExpressionExtensions.cs:
```diff
@@ -0,0 +1,83 @@
+                    return (expression as NuGetLicense).IsStandardLicense;
+
+                case LicenseExpressionType.Operator:
+
```
nit: remove extra line

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseExpressionExtensions.cs:
```diff
@@ -0,0 +1,83 @@
+        /// <summary>
+        /// A leaf node in an expression can only be a License or an Exception. Run a func on each one.
+        /// </summary>
+        /// <param name="expression"></param>
```
fill out `<param` tags or remove them if they do not provide useful information.

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseExpressionParser.cs:
```diff
@@ -0,0 +1,201 @@
+        /// Later the postfix expression is evaluated into an object model that represents the expression. Note that brackets are dropped in this conversion and this is not round-trippable.
+        /// The token precedence helps make sure that the expression is a valid infix one. 
+        /// </summary>
+        /// <param name="expression"></param>
```
fill out `<param` tag or remove it if it does not provide useful information.

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseExpressionParser.cs:
```diff
@@ -0,0 +1,201 @@
+        /// The token precedence helps make sure that the expression is a valid infix one. 
+        /// </summary>
+        /// <param name="expression"></param>
+        /// <returns>NuGetLicenseExpression</returns>
```
add more detailed information to `returns` tag. No need to have the type in comments

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseExpressionParser.cs:
```diff
@@ -0,0 +1,201 @@
+        /// <summary>
+        /// Tokenizes the expression as per the license expression rules. Throws if the string contains invalid characters.
+        /// </summary>
+        /// <param name="expression"></param>
```
fill out `<param` tag or remove it if it does not provide useful information.

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Licenses/NuGetLicenseExpressionParser.cs:
```diff
@@ -0,0 +1,201 @@
+        /// Tokenizes the expression as per the license expression rules. Throws if the string contains invalid characters.
+        /// </summary>
+        /// <param name="expression"></param>
+        /// <returns></returns>
```
fill out `returns` tag or remove empty

PatoBeltran commented on src/NuGet.Core/NuGet.Packaging/Properties/AssemblyInfo.cs:
```diff
@@ -1,6 +1,15 @@
 [assembly: CLSCompliant(true)]
+
+#if SIGNED_BUILD
+[assembly: InternalsVisibleTo("NuGet.Packaging.Test, PublicKey=0024000004800000940000000602000000240000525341310004000001000100b5fc90e7027f67871e773a8fde8938c81dd402ba65b9201d60593e96c492651e889cc13f1415ebb53fac1131ae0bd333c5ee6021672d9718ea31a8aebd0da0072f25d87dba6fc90ffd598ed4da35e44c398c454307e8e33b8426143daec9f596836f97c8f74750e5975c64e2189f45def46b2a2b1247adc3652bf5c308055da9")]
```
💯

PatoBeltran reviewed (approved):
Overall it looks good! Make sure to add comments to all public APIs and to have all comments without empty tags and with useful information.

nkolev92 committed:
Add style feedback

nkolev92 committed:
address feedback

nkolev92 committed:
Add the parse method to the abstract NuGetLicenseExpression class. make the parser internal

nkolev92 committed:
Add parsiong exception

nkolev92 committed:
add missing exception file

nkolev92 commented:
🔔 
Addressed the style comments.
Everythign that this method throws are NuGetLicenseExpressionParsingException. 
Made all the parsing methods on the license/exception internal.
Made the Parser internal. 
The only place to invoke the parsing now is by calling the static method on the NuGetLicenseExpression class (similar to VersionRange).

joelverhagen commented on test/NuGet.Core.Tests/NuGet.Packaging.Test/LicensesTests/NuGetLicenseExpressionParserTests.cs:
```diff
@@ -0,0 +1,211 @@
+        [Theory]
+        [InlineData("MIt WITH Classpath-exception-2.0", false, true)]
+        [InlineData("MIt AND BSD-2-Clause", false, true)]
+        [InlineData("MIt OR BSD-2-Clause", false, true)]
```
test `+` case too

joelverhagen replied:
Seems like you need a `true, false` and a `false, false` case too

joelverhagen replied:
Also, seems like we need some tests for non-standard exceptions. Perhaps I missed them but I don't see any specific tests for that.

nkolev92 replied:
The SPDX specification does not declare non-standard exception afaik.

joelverhagen commented on test/NuGet.Core.Tests/NuGet.Packaging.Test/LicensesTests/NuGetLicenseExpressionParserTests.cs:
```diff
@@ -0,0 +1,211 @@
+        [InlineData("MIT OR LPL-1.0+", "OR", true)]
+        [InlineData("(AMDPLPA AND BSD-2-Clause)", "AND", true)]
+        [InlineData("((( (AMDPLPA) AND BSD-2-Clause)))", "AND", true)]
+        [InlineData("(And+) AND or", "AND", false)]
```
`AND AND AND` is valid but non-standard, right?

nkolev92 replied:
Not quite. The license expression per SPDX is case sensitive. 
AND => operator
And => identifier, which will obviously get parsed as a non standard license. 

Per the design however, With WITH wITH would be invalid, as I did not find the concept of user defined exceptions.

joelverhagen commented on test/NuGet.Core.Tests/NuGet.Packaging.Test/LicensesTests/NuGetLicenseExpressionParserTests.cs:
```diff
@@ -0,0 +1,211 @@
+        [InlineData("((MIT) (AND) (LPL-1.0))")]
+        [InlineData("MIT (AND LPL-1.0)")]
+        [InlineData("MIT () OR LPL-1.0")]
+        [InlineData("MIT OR GPL-1.0 (WITH 389-exception OR LPL-1.0)")]
```
Also worth testing:

`()` - only parens
`MIT WITH OR LPL-1.0` - consecutive operators
`MIT AND OR LPL-1.0` - consecutive logical operators
`MIT++` - multiple pluses
`MIT+ OR LPL-1.0+++`  - multiple pluses in complex expression
`(MIT OR LPL-1.0)+` - attempt at some kind of distributed property

nkolev92 replied:
Nice test cases :)
Added them.

joelverhagen commented on test/NuGet.Core.Tests/NuGet.Packaging.Test/LicensesTests/NuGetLicenseExpressionParserTests.cs:
```diff
@@ -0,0 +1,211 @@
+        [Theory]
+        [InlineData("MIT OR LPL-1.0", "OR", true)]
+        [InlineData("MIT AND LPL-1.0", "AND", true)]
+        [InlineData("MIT OR (AFL-1.1 WITH 389-exception)", "OR", true)]
```
Do you have tests that assert the whole tree? It looks like we are just looking at the parsing of the root.

Perhaps you could do this easily with a round-trippable `.ToString()`? Or just baked the tree into a bunch of asserts in a `[Fact]`

nkolev92 replied:
In parses complex expression I do the following: 
So it kind of covers that.
Assert.Equal(expression.Replace("(", "").Replace(")", "").Trim(), licenseExpression.ToString());

I will add a fact just for sanity.

joelverhagen commented on test/NuGet.Core.Tests/NuGet.Packaging.Test/LicensesTests/NuGetLicenseExpressionParserTests.cs:
```diff
@@ -0,0 +1,211 @@
+        [InlineData("((( (AMDPLPA) AND BSD-2-Clause)))", "AND", true)]
+        [InlineData("(And+) AND or", "AND", false)]
+        [InlineData("(AMDPLPA AND BSD-2-Clause) AND (MIT WITH 389-exception)", "AND", true)]
+        [InlineData("(AMDPLPA AND BSD-2-Clause) AND (MIT OR Apache-2.0)", "AND", true)]
```
We should either reject or test leading, trailing, or extra whitespace (double spaces, `\r`, `\n`, tabs)

nkolev92 replied:
I have random white space tests in the tokenizer but for some reason not in the parser tests themselves (which use the tokenizer). 

I will add them here as well.

joelverhagen reviewed (approved):
Just some comments on unit test coverage. Looks great!

nkolev92 committed:
add more tests

nkolev92 commented:
@joelverhagen 
I've added a bunch more tests per your suggestions

jainaashish commented on src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionTokenizer.cs:
```diff
@@ -0,0 +1,123 @@
+
+                IList<LicenseExpressionToken> tokensAfterValue = null;
+
+                while (processingToken.Length > 0 && (processingToken[0] == '(' || processingToken[0] == ')'))
```
I thought you were going to simplified it...

nkolev92 replied:
Tracking issue for that https://github.com/NuGet/Home/issues/7302
